ambry-store/src/main/java/com.github.ambry.store/HardDeleter.java

```diff
@@ -111,17 +113,30 @@ public void run() {
     try {
       while (running.get()) {
         try {
-          if (!hardDelete()) {
-            isCaughtUp = true;
+          if (paused.get()) {
```

Comment:
Yeah, this can and should be cleaned up a bit. Looks more complicated than it has to be.
To answer your first question about why `while` and not `if` - in general one should enclose `wait()` in a while loop and recheck the condition when woken up - to handle [spurious wakeups](https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#wait()).